crates/sui-adapter/src/programmable_transactions/execution.rs

```diff
@@ -1006,41 +997,50 @@ fn is_string_arg<E: fmt::Debug, S: StorageView<E>>(
         | Type::Signer
         | Type::Reference(_)
         | Type::MutableReference(_)
-        | Type::TyParam(_) => return Ok(None),
-        // TODO should we support Option of string?
-        Type::StructInstantiation(_, _) => return Ok(None),
+        | Type::TyParam(_) => Ok(None),
+        Type::StructInstantiation(idx, targs) => {
+            let Some(s) = context.session.get_struct_type(*idx) else {
+                invariant_violation!("Loaded struct not found")
+            };
+            let resolved_struct = get_struct_ident(&s);
+            // is option of a string
+            if resolved_struct == RESOLVED_STD_OPTION && targs.len() == 1 {
+                let info_opt = additional_validation_layout(context, &targs[0])?;
+                Ok(info_opt.map(|layout| SpecialArgumentLayout::Option(Box::new(layout))))
+            } else {
+                Ok(None)
+            }
+        }
         Type::Vector(inner) => {
-            let info_opt = is_string_arg(context, inner)?;
-            return Ok(info_opt.map(|(string_name, layout)| {
-                (string_name, MoveTypeLayout::Vector(Box::new(layout)))
-            }));
+            let info_opt = additional_validation_layout(context, inner)?;
+            Ok(info_opt.map(|layout| SpecialArgumentLayout::Vector(Box::new(layout))))
         }
-        Type::Struct(idx) => idx,
-    };
-    let Some(s) = context.session.get_struct_type(*idx) else {
-        invariant_violation!("Loaded struct not found")
-    };
-    let resolved_struct = get_struct_ident(&s);
-    let string_name = if resolved_struct == RESOLVED_ASCII_STR {
-        RESOLVED_ASCII_STR
-    } else if resolved_struct == RESOLVED_UTF8_STR {
-        RESOLVED_UTF8_STR
-    } else {
-        return Ok(None);
-    };
-    let layout = MoveTypeLayout::Struct(MoveStructLayout::Runtime(vec![MoveTypeLayout::Vector(
-        Box::new(MoveTypeLayout::U8),
-    )]));
-    Ok(Some((string_name, layout)))
+        Type::Struct(idx) => {
+            let Some(s) = context.session.get_struct_type(*idx) else {
+                invariant_violation!("Loaded struct not found")
+            };
+            let resolved_struct = get_struct_ident(&s);
+            let layout = if resolved_struct == RESOLVED_ASCII_STR {
+                Some(SpecialArgumentLayout::Ascii)
+            } else if resolved_struct == RESOLVED_UTF8_STR {
+                Some(SpecialArgumentLayout::UTF8)
+            } else {
+                None
+            };
+            Ok(layout)
+        }
+    }
+}
+
+/// Special enum for values that need additional validation, in other words
+/// There is validation to do on top of the BCS layout. Currently only needed for
+/// strings
+pub enum SpecialArgumentLayout {
+    Option(Box<SpecialArgumentLayout>),
+    Vector(Box<SpecialArgumentLayout>),
```

Comment:
How deeply nested can these get?

Reply:
Currently? It is bounded by the max depth of the types in the VM (but otherwise arbitrary)